content/blog/2022-04-26-Superoptimizer.md

```diff
@@ -0,0 +1,83 @@
+
+While test vectors are likely to eliminate many semantically different enumerated programs, some enumerated programs may still fall through the cracks. At this point, it is the responsibility of the operator of the Superoptimizer to manually ensure that the proposed program is equivalent to the original program. 
+
+## Applications
+
+The author proposes several scenarios where Superoptimizer can be applied. The simplest scenario is to use the Superoptimizer on small chunks of code, such as small functions. For instance, the author applies the Superoptimizer on small mathematical functions such as the absolute value function or maximum and minimum functions. On the assumption that these functions are heavily used, the application of Superoptimizer on these functions can yield great improvement. 
+
+The Superoptimizer can also be applied to larger functions, with the caveat that some clever work has to be done to allow the Superoptimizer to run in a feasible time frame. For instance, the author applied the Superoptimizer to optimize a `printf` function.  By applying the Superoptimizer to small chunks of the `printf` function, the function could be improved incrementally. In fact, using the prior approach seems even better when one has an idea of what types of code are more frequently used. Having a profiler guide where these small chunks of code can be superoptimized seems like one approach to use this technique. 
+
+Another way the Superoptimizer can be used is to identify sequences of instructions that can be used for pruning. In this use case, the Superoptimizer can enumerate sets of instruction sequences equivalent to a single instruction. When a sequence of instructions is found to be in that set, we know that it is redundant to the original single instruction, and thus can be pruned. 
+
+Finally, the author also gives an instance of using the Superoptimizer to discover equivalent statements in computer architecture. As some instruction set architectures contain many equivalent instructions, it is possible to use the Superoptimizer to search for an equivalent sequence of simpler instructions that are equivalent to a more complex instruction in the ISA. 
+
+## Related Works
+
+The Superoptimizer has several limitations which influence how the Superoptimizer can be applied. One of the major limitations of Massalin’s work is that Superoptimizer takes a long time to run and find optimal programs. Another downside is that Superoptimizer only works on the Motorola 68000 and Intel 8086 architectures. Finally, yet another issue with Superoptimizer is that enumerated programs are not guaranteed to be equivalent to the original program. These downsides influence the design of future projects that build upon the original Superoptimizer. 
```

Comment:
Also, I'm not 100% sure we should count the target architecture support as a limitation… yes, technically it is limited in the sense that there are other architectures that exist in the world, but supporting _all_ architectures would be impossible, right? Maybe a different way to put it would be to just say that she focused on a couple of ISAs, and subsequent work has done the obvious thing of increasing target support.

Reply:
I also removed the target architecture limitation. I agree it does not make sense to support all architectures.